fix dialyzer warnings

hmmr opened the pull request.

Commit:
move code to be shared between riak_kv_{wm,pb}_timeseries, to _ts_util

Commit:
http client for TS

Enable the following requests:

GET    /ts/1/table/Table          single-key get
DELETE /ts/1/table/Table          single-key delete
PUT    /ts/1/table/Table          batch put
GET    /ts/1/table/Table/keys     list_keys
GET    /ts/1/coverage             coverage for a query
POST   /ts/1/query                execute SQL query

Additional parameters (keys, data, query) are to be supplied in
request body, as a JSON object.

Commit:
riak_kv_qry: filter out empty records

Commit:
Create riak_ts-develop-1.3

Commit:
detect and deal with {error,no_type} (table not acivated case)

Complementary to 9b62335 by atill

The call to claimant to determine a table's bucket type status is an
expensive operation.  It was introduced in a1c1e2e6, with the purpose to
detect and report the error condition where a TS operation is attempted on
a non-activated bucket type, but found to cause serious performance
degradation. Andy's commit removed the expensive call; this commit captures
and reports the error condition in various places in
riak_kv_{wm,pb}_timeseries and riak_kv_ts_util.

Commit:
make TS query served from own WM resource

andalso, check detailed per-table permissions when security is on.

Commit:
ignore non-relevant parameters in json for TS calls

Commit:
repin bitcask, other deps back to riak_ts-develop

because there were no changes for them to progress to 1.3, whereby
they lose the version and become riak_ts-develop.

Commit:
filter empty records at collection site in riak_kv_qry_worker

This improves on 5c56530.

Commit:
wm_timeseries: also support keys in url for get & delete

Commit:
wm_timeseries: disable key-fields-in-json-body mode for TS get and delete

The version with both key-in-json and key-in-url is forked into own branch
feature-az-http_ts_api_with_keys_in_json_in_body.

If a decision is taken to resurrect that branch, there is a complementary
branch, of the same name, in riak-erlang-http-client, which can produce
requests with key in json.

Commit:
wm_timeseries: use POST for batch puts

Commit:
collect internal TS API functions into riak_kv_ts_api

Commit:
cumulative update including

* don't support coverage API call in HTTP callbacks;
* unify permission strings, for shared use between wm and pb callbacks;
* straighten up function call chain in riak_kv_wm_timeseries, now that
  all api calls are mapped 1:1 onto HTTP methods;
* in particular, use wrq:path_info/2 to get table and keys, instead of
  clunky tokenization where it's no longer necessary;
* introduce security and validation checks for listkeys.

Commit:
dispatch all api_call branches per method, from one validate_request fun

Also, move around the check for no_such_table, and keep Mod and DDL once
found.

Commit:
avoid case clauses in favour of function clauses

Commit:
encode data for batch put as is, not at "data" key in a proplist

Commit:
wm_timeseries: simplify batch put data extraction from json

Commit:
WIP - Total rewrite of riak_kv_wm_timeseries

* Full usage of webmachine callbacks.

Add case for undefined SecContext

Many small fixes. Basic functionality works. Need to add more tests.

Changed ddl_fields/1 to ddl_fieldsand_types in prep for moving it to the DDL helper module.

Code re-org. Moved all helper modules to the end of the file.

Remove unused macro.

Move common functions to riak_kv_wm_ts_util

Use common authorize function in riak_kv_wm_timeseries and riak_kv_wm_timeserise_listkeys

Initial commit of riak_kv_wm_ts_util

riak_kv_wm_timeseries_listkey refactorings

* removed malformed_request
* changed CB_RV_SPEC to cb_rv_spec(T) type
* simplified forbidden/2
* removed check_permissions/2 as it is no longer needed (is_authorized/2 handles that job now)
* removed utf8_to_binary/1 as it is no longer used (riak_kv_wm_ts_util has it for those functions who needs it).

Moved table_module_exists to util module. Simpler resource_exists for riak_kv_wm_timeseries_listkeys.

rtimeseries_listkeys service available simplified and functions removed.

* error_out, handle_error and flat_format removed as they are no longer used.

moved local_key(Mod) to wm_ts_util

timeseries_listkeys refactored and produces html

Uniform disptach routes for timeseries in riak_kv_web

WIP. TS query now works for create table

WIP. timeseries describe query now works.

timeseries select query works.

riak_kv_vm_timeseries remove dead code

Trying to humour dialyzer, but is not yielding fully.

riak_kv_wm_timeseries_query reorganised to have the callbacks in the right order.

And all helper functions at the end of the module.

Comments in the top of the files aligned with what the code does.

Fix: timeseries_query with correct SQL for non-existing table now returns 404 instead of crashing.

timeseries_query: remove prefix, riak from #ctx. Use init:get_status() to judge availability.

timeseries_query: only POST allowed.

timeseries_query: removed debugging with checkpoint/1

timeseries_query: unsupported SQL type now leads to a 503

timeseries_query: query string now in body as text/plain

timeseries_query: attempting to create an existing table now return a 409

wm_timeseries: consistent throw structure.

WIP wm_timeseries_query: create and describe return json body.

wm_timeseries_util: set_error_message now produces a JSON body.

wm_timeseries: conversion to returning JSON bodies.

ts_query: Better handling of wrong queries.

wm_timeseries: fixed URL in comments.

ts listkeys: modified the return type to be text/plain and returning urls separated by newline.

Commit:
Add DDL compiler versioning and recompilation

Commit:
Revert DETS primary key back to BucketType; simplify old entry cleanup

Commit:
Committed changes to handle separate put and query reqs for PB and TTB

Commit:
Forgot to copy the include file from my sandbox

Commit:
Changed riak_kv_pb_timeseries service to handle up to message 104

Commit:
Added header file needed for ttb messages

Commit:
Updated to handle default values for Request atoms in coverage requests, and in build_sql_record when not called in response to a request

Commit:
Removed tabs

Commit:
Sync dets to disk after each writing instead of waiting for default
time period

Commit:
Begin work on differentiating PB / TTB my msg code only

lukebakken committed:
Some cleanup.

Commit:
amend create_table code path to take into account WITH-props

Commit:
fix dialyzer warnings

lehoff commented on src/riak_kv_wm_timeseries_listkeys.erl:
```diff
@@ -165,14 +165,10 @@ ts_keys_to_body(EncodedKeys, Table, Mod) ->
-    %%                          format_url(BaseUrl, KeyTypes, Key)
-    %%                  end,
-    %%                  Keys),
+    URLs = lists:map(fun(Key) ->
```
Please leave in a comment on dialyzer, with a `@todo` comment and the code that dialyzer does not accept.
Otherwise we will forget it.

Commit:
Split into PB and TTB services for TS

Commit:
Correctly encode an empty tsqueryresp

Commit:
Move decoding of tsputreq rows into riak_pb

Commit:
Move decoding of tsgetreq key into riak_pb

Commit:
Fix comment

erikleitch committed:
Merge pull request #1379 from basho/perf/query_ttb_encoding_lrb

READY: differentiate PB / TTB by msg code only

Commit:
Changed riak_kv_ttb_ts to use riak_ttb_codec:decode/1

Commit:
Reverted rebar.config, checked for tabs

Commit:
Fix badmatch error, whitespace

Commit:
Rename SQL "table" with DDL "record" naming for clarity

Commit:
Update tools.mk to generate correct regex for dialyzer

Commit:
Dialyzer fixes

Commit:
Maintain old DETS table of DDL entries to allow for downgrade

Commit:
wm_timeseries_listkeys: urlencode URLs^W keys

Commit:
wm_timeseries_listkeys: deal with sint64 and double, too

Commit:
Ensure legacy and current DETS table are, in fact, different

Commit:
Remove deleted rows from the result set at the coordinator level, this avoids checks through the selection funs and in pb.

Commit:
Handle the case where old DDL definition is undefined

Commit:
Move compiler version to riak_ql_ddl_compiler

borshop committed:
Merge pull request #1378 from basho/perf/query_ttb_encoding

READY: Perf/query ttb encoding

Reviewed-by: alexmoore

Commit:
Move riak_ql_ddl_compiler:compiler_version to
riak_ql_component:component_version

Commit:
have ts_util:queried_table/1 extract Table from a DDL

Commit:
wm_timeseries: cumulative commit to fix the following:

* accurately check and report missing/extra/unpaired field/value pairs
  in get/delete URLs;
* avoid raw Erlang terms in error messages;
* in particular, make lexer/parser error readable again;
* uniformly report error messages as text/plain;
* re-enable check for api_version;
* correctly check URL parameters (like ?timeout=5);
* correctly check for service availability (use init:get_status/0);
* use spaces around '=' etc, and reindent.

Commit:
Merge branch 'riak_ts-develop' of github.com:basho/riak_kv into at-query-deleted-rows

borshop committed:
Merge pull request #1383 from basho/at-query-deleted-rows

TS filter deleted rows

Reviewed-by: macintux

Commit:
Update a few comments based on code reviews

Commit:
Merge pull request #1377 from javajolt/bch-register-ddl-version

RTS-980,RTS-981,RTS-982: Recompile DDL

Commit:
Added branch back again...

Commit:
post-rebase fixups

Commit:
whitespace and indents

Commit:
move INSERT query code into riak_kv_qry

Previously, calling riak_kv_qry:submit("INSERT ...") would have no effect
because the actual code repackaging data and calling riak_kv_api:put_data
used to be in riak_kv_pb_timeseries (now riak_kv_pb_svc).

Incidentally, rename some query-type-specific callbacks in riak_kv_qry,
make do_describe return the common 3-tuple ({ColNames, ColTypes, Rows}).

Commit:
move some types from riak_kv_ts.hrl to riak_kv_pb_svc.erl, and export

Commit:
move PB error code defines to riak_kv_ts_svc (unused elsewhere)

Commit:
riak_kv_ts_svc: uniformly rename make_*_resp helper funs

Commit:
Merge pull request #1384 from basho/bugfix/add_back_row_encoding

Added back row encoding

Commit:
don't argue with dialyzer, esp. when it smells matching on records

Commit:
ts_api:query now returns uniform {ColNs, ColTs, Rows} for describe, too

Commit:
don't bother exporting ddl bucket type property for non-TS clients

Commit:
more accurately handle lexer errors

Commit:
Ensure rows are encoded for tsgetresp

borshop committed:
Merge pull request #1385 from basho/fixes/lrb/tsgetresp-encoding

Ensure rows are encoded for tsgetresp

Reviewed-by: hmmr

Commit:
ensure #rpberrorresp.msg only carries binaries, not strings

Commit:
Removed erroneous call to decode_query with tscoveragereq as second argument, and replaced it with undefined

borshop committed:
Merge pull request #1386 from basho/bugfix/repair_sub_tscoveragereq

Removed erroneous call to decode_query with tscoveragereq as second a…

Reviewed-by: lukebakken

Commit:
return [], not 'undefined' for NULL in DESCRIBE query

Commit:
Remove use of PB records for TTB responses

Commit:
properly assemble ts{query,get}resp wrt ColNames, ColTypes

Commit:
make DESRIBE query return compliant table

* rows as tuples, not lists of cells;
* [] for undefined.

Commit:
make DESRIBE query return compliant table

* rows as tuples, not lists of cells;
* [] for undefined.

Commit:
Merge branch 'riak_ts-develop' of github.com:basho/riak_kv into riak_ts-develop

Commit:
enforce binaries in #rpberrorresp.errmsg

Commit:
fix comments

Commit:
slightly optimise riak_kv_ts_util:validate_rows

Commit:
don't deal with #tscoverageresp internal structs in riak_kv

Move the relevant code to riak_pb_ts_codec. This will allow
riakc_ts:get_coverage to be served over TTB.

Commit:
Merge pull request #1388 from basho/features/lrb/simplify-ttb-further

Remove use of PB records for TTB responses

Commit:
Fix unit tests

Commit:
Merge pull request #1389 from basho/fix-az-ttb_friendly_outgoing_bits

make outgoing PB messages TTB-friendly

Commit:
Convert output lists to tuples for TTB; fix get notfound

Commit:
convert to tuples in lowest level

Commit:
fix eunit test

Commit:
Merge pull request #1391 from basho/fixes/ensure-result-rows-as-tuples

Query and get result rows as tuples

Commit:
make query returned table have rows as tuples

Commit:
Merge branch 'riak_ts-develop' of github.com:basho/riak_kv into riak_ts-develop

Commit:
Merge branch 'riak_ts-develop' into fixup-az-rebase_th_http_client_onto_develop

# Conflicts:
#	src/riak_kv_qry_worker.erl
#	src/riak_kv_ts_svc.erl
#	src/riak_kv_ttb_ts.erl

Commit:
qry_compiler: report too_many_subqueries with a binary description

This plays better with error presentation in riak_kv_ts_svc. Also,
treat this error as E_SUBMIT (1001), as it used to be.

Commit:
correctly report INSERT query errors

Commit:
query compiler errors should be treated as E_SUBMIT (1001)

This unbreaks some ts_* tests.

Commit:
do list_to_tuple after riak_kv_api:query, avoid re-converting for wm_ts

In riak_kv_wm_timeseries_query, for shipping to the client in json format,
records need to be lists. Let's avoid re-converting records, if only
for the HTTP API callbacks, and delegate the conversion to tuples to the
callers of riak_kv_api:query.

Commit:
ts_util: let validate_rows handle any()

This will conveniently detect non-tuples for records from the client.

Commit:
fix list_keys for short keys

It appears short keys, as in

  CREATE TABLE t (ts TIMESTAMP NOT NULL, PRIMARY KEY((ts),ts));

don't get sext-encoded, and are instead sent to the calling process
of riak_client:stream_list_keys as tuples, causing the resource streaming
fun to silently choke on sext:decode(Tuple). As a result, the http client
expecting chunked response, receives nothing.

This is probably a bug on the side of riak_kv, which we work around by
explicitly not attempting a sext:decode on tuples.

Commit:
fix tuple-list-tuple conversion-related eunit tests once again
